ios/engine/KMEI/KeymanEngine/Classes/Manager.swift

```diff
@@ -613,191 +613,97 @@ public class Manager: NSObject, UIGestureRecognizerDelegate {
-            } catch {
-              log.error("Could not create dir for download: \(error)")
-              throw KMPError.fileSystem
-            }
-            
-            var haveInstalledOne = false
-            for keyboard in installableKeyboards {
-              let storedPath = Storage.active.keyboardURL(for: keyboard)
-              
-              var installableFiles: [[Any]] = [["\(keyboardID).js", storedPath]]
-              if let osk = osk {
-                let oskPath = Storage.active.fontURL(forKeyboardID: keyboardID, filename: osk)
-                installableFiles.append([osk, oskPath])
-              }
-              
-              if let font = font {
-                let displayPath = Storage.active.fontURL(forKeyboardID: keyboardID, filename: font)
-                installableFiles.append([font, displayPath])
-              }
-              do {
-                for item in installableFiles {
-                  var filePath = folder
-                  if(FileManager.default.fileExists(atPath: (item[1] as! URL).path)) {
-                    try FileManager.default.removeItem(at: item[1] as! URL)
-                  }
-                  filePath.appendPathComponent(item[0] as! String)
-                  try FileManager.default.copyItem(at: filePath,
-                                                   to: item[1] as! URL)
-                  
-                }
-              } catch {
-                log.error("Error saving the download: \(error)")
-                throw KMPError.copyFiles
-              }
-              if !haveInstalledOne {
-                Manager.shared.addKeyboard(keyboard)
-                haveInstalledOne = true
-              }
+    guard let kmp = KeymanPackage.parse(folder) as? KeyboardKeymanPackage else {
+      throw KMPError.wrongPackageType
+    }
+
+    for k in kmp.keyboards {
+      let installableKeyboards : [InstallableKeyboard] = k.installableKeyboards
+      let keyboardID = k.keyboardId
+
+      do {
+        try FileManager.default.createDirectory(at: Storage.active.keyboardDir(forID: keyboardID),
+                                                withIntermediateDirectories: true)
+      } catch {
+        log.error("Could not create dir for download: \(error)")
+        throw KMPError.fileSystem
+      }
+
+      var haveInstalledOne = false
+      for keyboard in installableKeyboards {
+        let storedPath = Storage.active.keyboardURL(for: keyboard)
+
+        var installableFiles: [[Any]] = [["\(keyboardID).js", storedPath]]
+        if let osk = k.osk {
+          let oskPath = Storage.active.fontURL(forKeyboardID: keyboardID, filename: osk)
+          installableFiles.append([osk, oskPath])
+        }
+
+        if let font = k.font {
+          let displayPath = Storage.active.fontURL(forKeyboardID: keyboardID, filename: font)
+          installableFiles.append([font, displayPath])
+        }
+        do {
+          for item in installableFiles {
```

Comment:
Would it be better/possible to copy the entire folder?

Reply:
In another PR, yes.  This one is concerned with maintaining existing behavior while relying on a different parse source.